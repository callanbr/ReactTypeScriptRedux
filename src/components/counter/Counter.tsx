import { useState } from 'react';
import { useRootDispatch, useRootState } from '../../redux/hooks';
import {
  incrementOne,
  changeAmount,
  decrementOne,
} from '../../redux/features/counterSlice';

export const Counter = () => {
  const [inputValue, setInputValue] = useState<number>(0);
  const count = useRootState((state) => state.counter.count);
  const dispatch = useRootDispatch();

  const handleIncrementOne = () => dispatch(incrementOne());
  const handleDecrementOne = () => dispatch(decrementOne());
  const handleIncrementAmount = () => dispatch(changeAmount(inputValue));

  return (
    <div>
      <h2>{count}</h2>
      <button onClick={handleIncrementOne}>Increment One</button>
      <button onClick={handleDecrementOne}>Decrement One</button>
      <br />
      <input
        type="number"
        name="amountToChange"
        placeholder="Enter a number"
        onChange={(e) => setInputValue(parseInt(e.target.value) || 0)}
      />
      <button onClick={handleIncrementAmount}>Change Amount By</button>
    </div>
  );
};
