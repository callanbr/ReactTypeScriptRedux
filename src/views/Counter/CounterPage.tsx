// import { useSelector, useDispatch } from 'react-redux';
// import { RootState } from '../../redux/reducers';
// import { bindActionCreators } from 'redux';
// import { countActions } from '../../redux/actions';
import Counter from '../../components/counter/Counter';

export const ReduxPage = () => {
  // const count = useSelector((state: RootState) => state.count);
  // const dispatch = useDispatch();
  // const { increment, decrement } = bindActionCreators(countActions, dispatch);

  // const increment = (amount: number) => {
  //   dispatch({ type: 'INCREMENT', payload: amount });
  // };

  return (
    <div>
      <h1>Redux Page</h1>
      {/* <p>{count}</p>
      <button onClick={() => increment(1)}>Increment</button>
      <button onClick={() => decrement(1)}>Decrement</button> */}
      <Counter />
    </div>
  );
};
