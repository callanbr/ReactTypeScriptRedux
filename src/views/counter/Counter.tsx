import { Counter } from '../../components/counter/Counter';

export const CounterView = () => {
  return (
    <div>
      <h1>Counter Page</h1>
      <Counter />
    </div>
  );
};
