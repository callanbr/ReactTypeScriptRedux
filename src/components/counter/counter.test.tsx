import { configureStore } from '@reduxjs/toolkit';
import { render, screen, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';
import TestRenderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { Counter } from './Counter';
import counterReducer, {
  decrementOne,
  changeAmount,
  incrementOne,
} from '../../redux/features/counterSlice';
const { act } = TestRenderer;

const store = configureStore({
  reducer: {
    counter: counterReducer,
  },
});

afterEach(cleanup);

describe('renders correctly', () => {
  test('matches snapshot', () => {
    const tree = TestRenderer.create(
      <Provider store={store}>
        <Counter />
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });

  test('with initial count of 0', () => {
    render(
      <Provider store={store}>
        <Counter />
      </Provider>
    );
    const count = screen.getByRole('heading');
    expect(count).toHaveTextContent('0');
  });
});

describe('Redux', () => {
  test('initial state', () => {
    expect(store.getState()).toEqual({
      counter: {
        count: 0,
      },
    });
  });
  test('increment by 1', () => {
    act(() => {
      store.dispatch(incrementOne());
    });
    expect(store.getState()).toEqual({
      counter: {
        count: 1,
      },
    });
  });
  test('decrement by 1', () => {
    act(() => {
      store.dispatch(decrementOne());
    });
    expect(store.getState()).toEqual({
      counter: {
        count: 0,
      },
    });
  });
  test('change amount by -10', () => {
    act(() => {
      store.dispatch(changeAmount(-10));
    });
    expect(store.getState()).toEqual({
      counter: {
        count: -10,
      },
    });
  });
  test('change amount by +10', () => {
    act(() => {
      store.dispatch(changeAmount(10));
    });
    expect(store.getState()).toEqual({
      counter: {
        count: 0,
      },
    });
  });
});

describe('user clicked', () => {
  test('increment button', () => {
    act(() => {
      render(
        <Provider store={store}>
          <Counter />
        </Provider>
      );
      const incrementOne = screen.getByRole('button', {
        name: 'Increment One',
      });
      userEvent.click(incrementOne);
    });
    const count = screen.getByRole('heading');
    expect(count).toHaveTextContent('1');
  });
  test('decrement button', () => {
    act(() => {
      render(
        <Provider store={store}>
          <Counter />
        </Provider>
      );
      const decrementOne = screen.getByRole('button', {
        name: 'Decrement One',
      });
      userEvent.click(decrementOne);
    });
    const count = screen.getByRole('heading');
    expect(count).toHaveTextContent('0');
  });
  test('change amount by 100 from input', async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <Counter />
        </Provider>
      );
      const input = screen.getByRole('spinbutton');
      userEvent.type(input, '100');
      expect(input).toHaveValue(100);
      const changeAmount = screen.getByRole('button', {
        name: 'Change Amount By',
      });
      userEvent.click(changeAmount);
    });
    const count = screen.getByRole('heading');
    expect(count).toHaveTextContent('100');
  });
  test('change amount by -999 from input', async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <Counter />
        </Provider>
      );
      const input = screen.getByRole('spinbutton');
      userEvent.type(input, '-999');
      expect(input).toHaveValue(-999);
      const changeAmount = screen.getByRole('button', {
        name: 'Change Amount By',
      });
      userEvent.click(changeAmount);
    });
    const count = screen.getByRole('heading');
    expect(count).toHaveTextContent('-899');
  });
});
