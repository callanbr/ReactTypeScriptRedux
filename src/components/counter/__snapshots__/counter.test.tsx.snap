// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly matches snapshot 1`] = `
<div>
  <h2>
    0
  </h2>
  <button
    onClick={[Function]}
  >
    Increment One
  </button>
  <button
    onClick={[Function]}
  >
    Decrement One
  </button>
  <br />
  <input
    name="amountToChange"
    onChange={[Function]}
    placeholder="Enter a number"
    type="number"
  />
  <button
    onClick={[Function]}
  >
    Change Amount By
  </button>
</div>
`;
